### PR TITLE
Update unlinking a Slack app

### LINKED ISSUE #131: Issues with revoking the oAuth token
When attempting to unlink the Slack app, if it fails, the "Slack App Unlinked Successfully" message shows even if the app was not actually unlinked.

The `revoke_oauth_token` method does return data/false on success/failure, but the `delete_oauth_token` method is just [assuming that it's successful without checking the result](https://github.com/easydigitaldownloads/edd-slack/blob/master/core/ssl-only/class-edd-slack-app-oauth-settings.php#L376), and showing the admin notice. That line should be changed to a variable assignment, and if it's false, a different notice should be shown.

There also needs to be a way to indicate to the site owner that the connection is no longer working. There are multiple calls going to get the users list, and without authorization, they're failing, but continue to happen.

I'm also looking at that button and think maybe its behavior should be updated to include some kind of validation or nonce check, and maybe made an ajax related function instead of loading on `init` as it does now.

If the token/secret/etc is updated in Slack and the site's connection is broken, the site owner doesn't have any way of knowing that. The token is only deleted if the revocation is successful, but not if Slack returns something like `invalid_auth`.

### PR DESCRIPTION
**Pull request by robincornett**
Fixes #131 

Proposed changes:
1. Send the oAuth token as an `Authorization` header instead of appending to the URL.
2. Use `GET` instead of `POST` for revoking.
3. Restrict revoking to administrator level users, and update the logic in that function to be a little more readable.
4. When viewing the settings screen for the extension, send a quick authorization check to Slack to see if the token is still valid. If it's not, delete the setting.
5. Rework the logic for setting the oAuth token--a bit more verbose but easier to read for me.

### CODE REVIEW

**Comment by mihaijoldis** [comment]
While testing this I was able to constantly make it brek the connection to the App. Sometimes after several link/unlin attempts and navigating to a non EDD page so that the URL query params are reset the authentication process just reloads the EDD slack settings page.

Something inside the store_oauth_token() method is causing this as the code returns early here
```
		if ( 'saving' !== $state || ! $token_type ) {
			return false;
		}
```
The `$_GET` dump right above that when it fails looks like this
```
[12-May-2021 14:51:59 UTC] Start GET: Array
(
    [post_type] => download
    [page] => edd-settings
    [tab] => extensions
    [section] => edd-slack-settings
    [code] => 2056589097684.2057878671154.9cff8cb9ed31269c840cfcdb198fea5ffd0aa3c6cdf57fafb787b9624b4e8232
    [state] => 

)
```

**Comment by robincornett** [comment]
@mihaijoldis I've updated the linking behavior to add a redirect after checking in with slack so that the token is not...exposed, I guess, in the URL. Also added more checks to the token setting method. Can you check again and see if you can still break it? (Seriously I cannot.)

**Comment by mihaijoldis** [comment]
@robincornett managed to break on the second attempt to reconnect.
added the debug entries I had before and again it seems to return `if ( 'saving' !== $state || ! $token_type ) {`

**Comment by mihaijoldis** [comment]
Here is a small recording: https://www.loom.com/share/873c8597612743bfb8b3a938b5192de2
You can see that one the modal window closes the main settings page just reloads.

**Comment by robincornett** [comment]
@mihaijoldis sorry, can you check again? I just tried it on the testing site and it linked successfully for me. I adjusted the behavior so that the user should not see the code in the URL again--once it's been recorded, the site redirects to `https://edd.thepatternmall.com/wp-admin/edit.php?post_type=download&page=edd-settings&tab=extensions&section=edd-slack-settings&edd_slack_oauth=success&edd_slack_type=updated`.

**Comment by mihaijoldis** [comment]
@robincornett its deffinetly better now as I can constantly get it to connect fine. In a few tests I ran like 10-15 quick connect/disconnects it only broke once. The URL on the EDD it redirected to was

`https://edd.thepatternmall.com/wp-admin/edit.php?post_type=download&page=edd-settings&tab=extensions&section=edd-slack-settings&code=2056589097684.2088391303457.95719a88aced88a1e0aea594f2871ec9e4a4dfd0f4a18e267c5ea2acc5d23104&state`

File: core/ssl-only/class-edd-slack-app-oauth-settings.php [hunk 1]
```diff
@@ -269,148 +319,214 @@ public function add_slack_invites_oauth_register( $args ) {
...
+			if ( empty( $oauth_request->ok ) ) {
+				$redirect_uri = add_query_arg(
+					array(
+						'edd_slack_oauth' => 'fail',
+						'edd_slack_type'  => 'error',
+					),
+					$redirect_uri
+				);
+
+				wp_safe_redirect( $redirect_uri );
```
Comment (ashleyfae) [inline core/ssl-only/class-edd-slack-app-oauth-settings.php:399]:
```suggestion
				wp_safe_redirect( $redirect_uri );
				exit;
```

Without the exit, it actually keeps executing the rest of the code and you end up on the success page.

File: core/ssl-only/class-edd-slack-app-oauth-settings.php [hunk 2]
```diff
@@ -269,148 +319,214 @@ public function add_slack_invites_oauth_register( $args ) {
...
+				$granted_client_scope = edd_update_option( 'slack_app_has_client_scope', true );
+				$redirect_uri         = add_query_arg(
+					array(
+						'edd_slack_oauth' => 'teams',
+						'edd_slack_type'  => 'updated',
+					),
+					$redirect_uri
+				);
+			}
+			wp_safe_redirect( $redirect_uri );
```
Comment (ashleyfae) [inline core/ssl-only/class-edd-slack-app-oauth-settings.php:426]:
```suggestion
			wp_safe_redirect( $redirect_uri );
			exit;
```

File: core/ssl-only/class-edd-slack-app-oauth-settings.php [hunk 3]
```diff
@@ -187,52 +194,97 @@ public function add_oauth_registration_button( $args ) {
...
+				$slack_uri = add_query_arg(
+					array(
+						'client_id'    => $client_id,
+						'scope'        => $scope,
+						'redirect_uri' => $redirect_uri,
+					),
+					'https://slack.com/oauth/authorize'
+				);
+				?>
+				<a href="<?php echo $slack_uri; ?>" target="_self" class="edd-slack-app-auth button button-primary" data-token_type="main">
```
Comment (ashleyfae) [inline core/ssl-only/class-edd-slack-app-oauth-settings.php:224]:
```suggestion
				<a href="<?php echo esc_url( $slack_uri ); ?>" target="_self" class="edd-slack-app-auth button button-primary" data-token_type="main">
```

**Review by ashleyfae — CHANGES_REQUESTED** [review]
This is working for me! Just requested a couple improvements, then we'll be good to go.

**Review by ashleyfae — APPROVED** [review]

**Pull request merged**